### PR TITLE
Enable most SDL-required C/C++ warnings

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Explicitly enables most SDL-required C/C++ warnings.

Due to the work in https://github.com/dotnet/runtime/issues/66154 (thanks all!), we're already clean with respect to these warnings, so all this PR does is establish a new baseline and make BinSkim happy.

Warning **C4242** is newly tracked since the last time we performed this task. We are not clean with regard to this warning due to some hits in libunwind, so I've left it commented out for the time being. Will open a new issue to track compiling clean for that warning code specifically.

### CODE REVIEW

**Comment by GrabYourPitchforks** [comment]
I broke CI by triggering a manual run before sending this PR. Whoops. :)

**Comment by GrabYourPitchforks** [comment]
Force-pushed to abort both dueling CI runs and trigger a new clean run.

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -832,10 +832,31 @@ if (MSVC)
...
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4551>) # Function call missing argument list.
-  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4700>) # Local used w/o being initialized.
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4640>) # 'instance' : construction of local static object is not thread-safe
   add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4806>) # Unsafe operation involving type 'bool'.
 
+  # SDL requires the below warnings to be treated as errors:
+  # (Search Liquid for requirement Microsoft.Security.SystemsADM.10086)
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4055>) # 'conversion' : from data pointer 'type1' to function pointer 'type2'
+  add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4146>) # unary minus operator applied to unsigned type, result still unsigned
+  # add_compile_options($<$<COMPILE_LANGUAGE:C,CXX>:/we4242>) # 'identifier' : conversion from 'type1' to 'type2', possible loss of data
```
Comment (jkotas) [inline eng/native/configurecompiler.cmake:842]: You can enable it here, and disable it just for libunwind that has the violations

https://github.com/dotnet/runtime/blob/main/src/native/external/libunwind_extras/CMakeLists.txt#L143

**Comment by GrabYourPitchforks** [comment]
I searched for occurrences of `/wdxxxx` and `-wdxxxx` for the SDL-required C/C++ warning numbers and didn't see any matches beyond the file Jan called out above.

**Review by jkotas — APPROVED** [review]
Thanks

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Review by AaronRobinsonMSFT — APPROVED** [review]
Yay!

**Review by hoyosjs — APPROVED** [review]

**Pull request merged**